dre-service/src/main/java/io/cdap/re/RulesDB.java

```diff
@@ -147,376 +199,399 @@ public void updateRule(String id, RuleRequest rule) throws RuleNotFoundException
+    rulebookTable.upsert(updatedRulebookFields);
+  }
 
-    long version = row.getLong(VERSION);
-    version = version + 1;
-    byte[][] values = new byte[][] {
-      Bytes.toBytes(rb.getDescription()),
-      Bytes.toBytes(rb.getUser()),
-      Bytes.toBytes(rb.getSource()),
-      Bytes.toBytes(version),
-      Bytes.toBytes(rb.getRulesString())
-    };
+  void addRuleToRulebook(String rulebookId, String ruleId)
+    throws RuleAlreadyExistsException, RulebookNotFoundException, RuleNotFoundException, IOException {
+    Collection<Field<?>> rulebookKeyFields = new ArrayList<>();
+    rulebookKeyFields.add(Fields.stringField(ID_COL, rulebookId));
 
-    rulebook.put(toKey(rb.getId()), columns, values);
-  }
+    Optional<StructuredRow> optionalRulebookStructuredRow = rulebookTable.read(rulebookKeyFields);
 
-  public void addRuleToRulebook(String rulebookId, String ruleId) throws RuleAlreadyExistsException,
-    RulebookNotFoundException, RuleNotFoundException {
-    Row row = rulebook.get(toKey(rulebookId));
-    if (row.isEmpty()) {
+    if (!optionalRulebookStructuredRow.isPresent()) {
       throw new RulebookNotFoundException(
         String.format("Rulebook '%s' not found.", rulebookId)
       );
     }
 
-    if (rules.get(toKey(ruleId)).isEmpty()) {
+    Collection<Field<?>> ruleKeyFields = new ArrayList<>();
+    ruleKeyFields.add(Fields.stringField(ID_COL, ruleId));
+
+    Optional<StructuredRow> optionalRuleStructuredRow = rulesTable.read(ruleKeyFields);
+
+    if (!optionalRuleStructuredRow.isPresent()) {
       throw new RuleNotFoundException(
         String.format("Attempt to add rule '%s' to rulebook '%s' failed as rule doesn't exist in rules database.",
                       rulebookId, ruleId)
       );
     }
 
-    long version = row.getLong(VERSION);
-    List<String> rules = convertRulesToSet(row.getString(RULES));
+    StructuredRow rulebookRow = optionalRulebookStructuredRow.get();
+    List<String> rules = convertRulesToSet(rulebookRow.getString(RULES_COL));
+
     if (rules.contains(ruleId)) {
       throw new RuleAlreadyExistsException(
         String.format("Rule '%s' already exists in the rulebook '%s'.", ruleId, rulebookId)
       );
     }
+
     rules.add(ruleId);
 
-    byte[][] columns = new byte[][] {
-      UPDATED, VERSION, RULES
-    };
-
-    version = version + 1;
-    byte[][] values = new byte[][] {
-      Bytes.toBytes(getCurrentTime()),
-      Bytes.toBytes(version),
-      Bytes.toBytes(Joiner.on(",").join(rules))
-    };
-    rulebook.put(toKey(rulebookId), columns, values);
+    Collection<Field<?>> rulebookFields = new ArrayList<>();
+    rulebookFields.add(Fields.stringField(ID_COL, rulebookId));
+    rulebookFields.add(Fields.longField(UPDATED_COL, getCurrentTime()));
+    rulebookFields.add(Fields.longField(VERSION_COL, rulebookRow.getLong(VERSION_COL)));
+    rulebookFields.add(Fields.stringField(RULES_COL, GSON.toJson(Joiner.on(",").join(rules))));
```

Comment:
Don't the other methods convert the rules directly to json? Or is RulebookRequest.getRules() returning a String and not a List?

The same data format should be used everywhere (a json array is better than joining on a comma).

Reply:
`rules` is a `List<String>`. Are you saying that I should do `GSON.toJson(rules)`?

Reply:
yes, the methods above to Gson.toJson(rulebookRequest.getRules())

Reply:
Yes, they are `rulebookRequest.getRules()` returns a `List<String>`. I will change the joiner with GSON.